src/dbnode/integration/setup.go

```diff
@@ -574,6 +549,96 @@ func (ts *testSetup) sleepFor10xTickMinimumInterval() {
 	time.Sleep(opts.TickMinimumInterval() * 10)
 }
 
+func (ts *testSetup) httpClusterAddr() string {
+	if addr := ts.opts.HTTPClusterAddr(); addr != "" {
+		return addr
+	}
+	return *httpClusterAddr
+}
+
+func (ts *testSetup) httpNodeAddr() string {
+	if addr := ts.opts.HTTPNodeAddr(); addr != "" {
+		return addr
+	}
+	return *httpNodeAddr
+}
+
+func (ts *testSetup) tchannelClusterAddr() string {
+	if addr := ts.opts.TChannelClusterAddr(); addr != "" {
+		return addr
+	}
+	return *tchannelClusterAddr
+}
+
+func (ts *testSetup) tchannelNodeAddr() string {
+	if addr := ts.opts.TChannelNodeAddr(); addr != "" {
+		return addr
+	}
+	return *tchannelNodeAddr
+}
+
+func (ts *testSetup) httpDebugAddr() string {
+	if addr := ts.opts.HTTPDebugAddr(); addr != "" {
+		return addr
+	}
+	return *httpDebugAddr
+}
+
+func (ts *testSetup) maybeResetClients() error {
+	if ts.m3dbClient == nil {
+		// Recreate the clients as their session was destroyed by stopServer()
+		adminClient, verificationAdminClient, err := newClients(
+			ts.topoInit, ts.opts, ts.hostID, ts.tchannelNodeAddr())
+		if err != nil {
+			return err
+		}
+		ts.m3dbClient = adminClient.(client.Client)
+		ts.m3dbAdminClient = adminClient
+		ts.m3dbVerificationAdminClient = verificationAdminClient
+	}
+
+	return nil
+}
+
+func newClients(
+	topoInit topology.Initializer,
+	opts testOptions,
+	id,
+	tchannelNodeAddr string,
+) (client.AdminClient, client.AdminClient, error) {
+	var (
+		clientOpts = defaultClientOptions(topoInit).
+				SetClusterConnectTimeout(opts.ClusterConnectionTimeout()).
+				SetWriteConsistencyLevel(opts.WriteConsistencyLevel()).
+				SetTopologyInitializer(topoInit)
+
+		// We need two distinct clients where one has the origin set to the same ID as the
```

Comment:
+1 for the explanation. How would you feel about moving it to the struct definition

Reply:
sure